### ISSUE TITLE
[BUG] filesystem ID check is bypassed when disk name is specified

### ISSUE THREAD

**Issue by boknowswiki** [longhorn/longhorn#1285]
**Describe the bug**
This is to make sure there is no two or more disk paths are in the same filesystem.
To avoid disk usage calculation issue.

This issue was found during testing [#1219 ](https://github.com/longhorn/longhorn/issues/1219)

**To Reproduce**
1. The Longhorn node contains no disk.
2. Go to Longhorn-UI "Setting/General/".
3. Click on "Create Default Disk on Labeled Nodes" to enable it.
4. The Kubernetes node object of the same name contains the label:
     `node.longhorn.io/create-default-disk: 'config' `
     and an annotation node.longhorn.io/default-disks-config, for example:
     ```
    node.longhorn.io/default-disks-config: 
    '[{"path":"/var","allowScheduling":true, "name":"name1"},
    {"path":"/home","allowScheduling":true,"storageReserved":1024,
    "tags":["ssd","fast"], "name":"name2"}]'
    ```
4. See two disks got created, but shows "Unschedulable" in the status section.

**Expected behavior**
No disk should be created and there should be error message about disk /home is the same filesystem with /var, fsid <FSID Number>"

**Log**
If applicable, add the Longhorn managers' log when the issue happens.

"time="2020-05-01T23:09:19Z" level=debug msg="Start backup store monitoring for s3://backupbucket@us-east-1/"
time="2020-05-01T23:09:22Z" level=info msg="Event(v1.ObjectReference{Kind:\"Node\", Namespace:\"longhorn-system\", Name:\"botao-linodecluster-dev-1\", UID:\"c27d4de5-b494-41e7-9820-68822687f13c\", APIVersion:\"longhorn.io/v1beta1\", ResourceVersion:\"383787\", FieldPath:\"\"}): type: 'Normal' reason: 'Ready' Node botao-linodecluster-dev-1 is ready"
time="2020-05-01T23:09:22Z" level=debug msg="Requeue longhorn-system/botao-linodecluster-dev-1 due to conflict: Operation cannot be fulfilled on nodes.longhorn.io \"botao-linodecluster-dev-1\": the object has been modified; please apply your changes to the latest version and try again"
time="2020-05-01T23:18:50Z" level=info msg="Event(v1.ObjectReference{Kind:\"Node\", Namespace:\"longhorn-system\", Name:\"botao-linodecluster-dev-1\", UID:\"c27d4de5-b494-41e7-9820-68822687f13c\", APIVersion:\"longhorn.io/v1beta1\", ResourceVersion:\"386141\", FieldPath:\"\"}): type: 'Warning' reason: 'Ready' Disk linode(/var) on node botao-linodecluster-dev-1 is not ready: disk has same file system ID 58fe937c58377e45 as other disks [linode root]"
time="2020-05-01T23:18:50Z" level=info msg="Event(v1.ObjectReference{Kind:\"Node\", Namespace:\"longhorn-system\", Name:\"botao-linodecluster-dev-1\", UID:\"c27d4de5-b494-41e7-9820-68822687f13c\", APIVersion:\"longhorn.io/v1beta1\", ResourceVersion:\"386141\", FieldPath:\"\"}): type: 'Warning' reason: 'Ready' Disk root(/home) on node botao-linodecluster-dev-1 is not ready: disk has same file system ID 58fe937c58377e45 as other disks [linode root]"
time="2020-05-01T23:18:50Z" level=info msg="Event(v1.ObjectReference{Kind:\"Node\", Namespace:\"longhorn-system\", Name:\"botao-linodecluster-dev-1\", UID:\"c27d4de5-b494-41e7-9820-68822687f13c\", APIVersion:\"longhorn.io/v1beta1\", ResourceVersion:\"386141\", FieldPath:\"\"}): type: 'Warning' reason: 'Schedulable' the disk linode(/var) on the node botao-linodecluster-dev-1 has 0 available, but requires reserved 1024, minimal 10% to schedule more replicas"

**Environment:**
 - Longhorn version:v0.8.1
 - Kubernetes version:v1.17.4
 - Node OS type and version:Ubuntu18.04 LTS

**Additional context**
Reference document:
https://github.com/longhorn/longhorn/blob/master/enhancements/20200331-replace-filesystem-id-key-in-disk-map.md

**Comment by boknowswiki** [comment]
The test case for this issue has been merge at:

https://github.com/longhorn/longhorn-tests/pull/308

**Comment by khushboo-rancher** [comment]
Verified on master - 05-08-2020
Verification - **Passed**

More than one disk path can't be created on same file system.
logs:
```
time="2020-05-11T19:31:19Z" level=warning msg="Kubernetes node: invalid annotation node.longhorn.io/default-disks-config: config: the disk /home is the samefile system with /var, fsid 4b0d97e476dc21d3"
time="2020-05-11T19:31:29Z" level=warning msg="Kubernetes node: invalid annotation node.longhorn.io/default-disks-config: config: the disk /home is the samefile system with /var, fsid 4b0d97e476dc21d3"
time="2020-05-11T19:31:29Z" level=warning msg="Kubernetes node: invalid annotation node.longhorn.io/default-disks-config: config: the disk /home is the samefile system with /var, fsid 4b0d97e476dc21d3"
time="2020-05-11T19:31:59Z" level=warning msg="Kubernetes node: invalid annotation node.longhorn.io/default-disks-config: config: the disk /home is the samefile system with /var, fsid 4b0d97e476dc21d3"
```

**Comment by meldafrawi** [comment]
Manual validation **Passed** by @khushboo-rancher 
Integration test implemented by @boknowswiki & merged.
Integration test passed in Nightly integration test.

**Comment by samos667** [comment]
It's intended to be applied on block device too ? Using this default config:

```json
[
  {
    "path": "/dev/disk/by-id/nvme-eui.002538d3***",
    "diskType": "block",
    "allowScheduling": true,
    "storageReserved": 0,
    "evictionRequested": false,
    "tags": [
      "ssd",
      "fast"
    ]
  },
  {
    "path": "/dev/disk/by-id/nvme-eui.002538bc***",
    "diskType": "block",
    "allowScheduling": true,
    "storageReserved": 0,
    "evictionRequested": false,
    "tags": [
      "ssd",
      "fast"
    ]
  }
]
```

Throw me this error:

```
error="the disk /dev/disk/by-id/nvme-eui.002538d3*** is the samefile system with /dev/disk/by-id/nvme-eui.002538bc***, diskID 4f490cc68acbeedd"
```

Edit:
Using longhorn node CR declaration:

```yaml
  disks:
    ssd1:
      allowScheduling: true
      diskDriver: auto
      diskType: block
      evictionRequested: false
      path: /dev/disk/by-id/nvme-eui.002538bc***
      storageReserved: 0
      tags: []
    ssd2:
      allowScheduling: true
      diskDriver: auto
      diskType: block
      evictionRequested: false
      path: /dev/disk/by-id/nvme-eui.002538d3***
      storageReserved: 0
      tags: []
```
Work perfectly. 

When testing the annotation declaration with only 1 disk this work (logic you will tell me) but I've noticed that `storageReserved: 0` was not applied (because defaulting to the 30% value), the CR node method don't have this issue. It seems like declaration via annotations is not really maintained/used or I'm wrong ?

**Comment by kolonelkrazy** [comment]
> When testing the annotation declaration with only 1 disk this work (logic you will tell me) but I've noticed that `storageReserved: 0` was not applied (because defaulting to the 30% value), the CR node method don't have this issue. It seems like declaration via annotations is not really maintained/used or I'm wrong ?

Same problem `storageReserved: 0` is not respected when it's an annotation on a node

**Comment by derekbit** [comment]
> > When testing the annotation declaration with only 1 disk this work (logic you will tell me) but I've noticed that `storageReserved: 0` was not applied (because defaulting to the 30% value), the CR node method don't have this issue. It seems like declaration via annotations is not really maintained/used or I'm wrong ?
> 
> Same problem `storageReserved: 0` is not respected when it's an annotation on a node

@kolonelkrazy Please follow up on https://github.com/longhorn/longhorn/issues/11543 and https://github.com/longhorn/longhorn/issues/11444. We will fix and improve it in v1.11.0.

cc @COLDTURNIP
